src/utilities/hooks.ts

```diff
@@ -0,0 +1,71 @@
+import { inject, onUnmounted } from 'vue'
+import { useRouterDepth } from '@/compositions/useRouterDepth'
+import { RouterNotInstalledError } from '@/errors/routerNotInstalledError'
+import { RouteMiddleware } from '@/types'
+import { asArray } from '@/utilities'
+import { AddRouteHook, RouteHookRemove, RouteHookType, addRouteHookInjectionKey } from '@/utilities/createRouterHooks'
+import { RouterRoute } from '@/utilities/createRouterRoute'
+
+/* Desired Hooks
+beforeRouteEnter
+beforeRouteUpdate
+beforeRouteLeave
+afterRouteEnter
+afterRouteUpdate
+afterRouteLeave
+*/
+
+type ComponentHookCondition = (to: RouterRoute, from: RouterRoute | null, depth: number) => boolean
+
+function factory(type: RouteHookType, condition: ComponentHookCondition): AddRouteHook {
+  return (middleware) => {
+    const depth = useRouterDepth()
+    const addRouteHook = inject(addRouteHookInjectionKey)
+
+    if (!addRouteHook) {
+      throw new RouterNotInstalledError()
+    }
+
+    const remove = asArray(middleware).map(middleware => {
+      const hook: RouteMiddleware = (to, context) => {
+        if (!condition(to, context.from, depth)) {
+          return
+        }
+
+        middleware(to, context)
+      }
+
+      return addRouteHook(type, hook)
+    })
+
+    const removeAll: RouteHookRemove = () => remove.forEach(fn => fn())
+
+    onUnmounted(removeAll)
+
+    return removeAll
+  }
+}
+
+const isComponentEnter: ComponentHookCondition = (to, from, depth) => {
```

Comment:
Starting here you can see the lifecycle conditions for component level hooks. I think these same conditions will be used for route level hooks.